Add runtime metrics in `System.Diagnostics.DiagnosticSource`

stevejgordon opened the pull request:
Fixes #85372

This adds new runtime metrics to align with the proposed [semantic conventions](https://github.com/open-telemetry/semantic-conventions/pull/1035). I am opening this as a draft until the naming is finalised in the OTel semantic conventions.

The metrics are essentially a port of the existing [OTel contrib runtime metrics](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.Runtime). Small changes have been made to optimise the performance and align the naming with those being proposed and discussed over in the semantic conventions repo.

cc @tarekgh and @noahfalk

stevejgordon committed:
Add runtime metrics

stevejgordon committed:
Add additional metrics

stevejgordon committed:
Add an initial test

stevejgordon committed:
Use GC.MaxGeneration

stevejgordon committed:
Make test less brittle

stevejgordon committed:
PR feedback and additional tests

cijothomas commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/RuntimeMetrics.cs:
```diff
@@ -0,0 +1,248 @@
+        {
+            AppDomain.CurrentDomain.FirstChanceException += (source, e) =>
+            {
+                if (t_handlingFirstChanceException) return;
```
could you add a comment on why this check is required?

stevejgordon replied:
Added a comment.

lmolkova commented:
@stevejgordon thanks a lot for doing this!

we synced offline and made some common suggestions on https://github.com/open-telemetry/semantic-conventions/pull/1035#issuecomment-2221544734. Hope it'd help you focus on this PR and get it merged 🤞

stevejgordon committed:
Add code comment

stevejgordon committed:
Update runtime metrics to align with latest semcon changes

stevejgordon commented:
@lmolkova, happy to contribute. Thanks for clarifying the suggestions in the semantic conventions. I've made those changes and aligned this PR with the updated naming.

stevejgordon committed:
Fix/update field names

tarekgh commented:
@carlossanlop @adamsitnik @ericstj I am seeing the errors `/__w/1/s/.packages/microsoft.dotnet.sharedframework.sdk/9.0.0-beta.24360.1/targets/sharedfx.targets(484,5): error : System.IO.Pipes > System.Net.Sockets > System.Net.NameResolution > System.Diagnostics.DiagnosticSource > System.Diagnostics.Process > System.IO.Pipes [/__w/1/s/src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj]` while this PR didn't add or change any dependencies. Any idea ideas why? the change started to call things like `Process process = Process.GetCurrentProcess();`.

tarekgh commented:
I talked offline with @ericstj and looks we introduced the dependency on System.Diagnostics.Process which caused the cyclic dependency. 

@noahfalk @samsp-msft @lmolkova @reyang do you see any issue if we move this feature to `Microsoft.Extensions.Diagnostics` library instead?

reyang commented:
> I talked offline with @ericstj and looks we introduced the dependency on System.Diagnostics.Process which caused the cyclic dependency.
> 
> @noahfalk @samsp-msft @lmolkova @reyang do you see any issue if we move this feature to `Microsoft.Extensions.Diagnostics` library instead?

No issue from what I can see, thanks!

stevejgordon commented:
> @noahfalk @samsp-msft @lmolkova @reyang do you see any issue if we move this feature to `Microsoft.Extensions.Diagnostics` library instead?

Unfortunately, there might be an issue with this. The implementation relies on the initialisation of the static meter from the ctor of `MeterListener`.  We'd need an alternative way to ensure the Meter is created.

tarekgh commented:
@jkotas we have the switch `System.Diagnostics.Metrics.Meter.IsSupported` to allow disabling the metrics. Is this can be acceptable mitigation for AOT if apps care about the size and not using metrics? I am just thinking loudly here. 

> Unfortunately, there might be an issue with this. The implementation relies on the initialisation of the static meter from the ctor of MeterListener. We'd need an alternative way to ensure the Meter is created.

Yes, I am trying to see if there are any other concerns that can prevent us moving it to the other library. For alternative way to create the Meter, I am thinking in the direction if we can keep creating the Meter in System.Diagnostics.DiagnosticSource but having instruments that are bringing the new dependencies be created from the extensions library. Of course we'll need to have a way to pick the meter instance from System.Diagnostics.DiagnosticSource. I am still thinking of other alternatives or viable solutions.

noahfalk commented:
> I think we either need to avoid taking large dependency closure in the implementation and then it is ok to enable it by default; or it needs to opt-in and then the large dependency closure is ok (not ideal though).

I would aim for the former - avoid taking large dependency closure. In theory runtime metrics should be nothing more than some calls into low-level runtime APIs (GC, JIT, assemblies, threadpool) + a tiny bit of data from the OS (some cpu and memory usage info)

tarekgh commented:
@noahfalk there are more dependencies and not just Process one. Yes, process one is what is causing cycles, but I am not sure if we are ok for the other dependencies. 

```Xml
    <Reference Include="System.Diagnostics.Process" />
    <Reference Include="System.Threading.ThreadPool" />
    <Reference Include="System.ComponentModel.Primitives" />
```

tarekgh commented:
> I'm hopping between meetings and haven't had a chance to look more closely, but could we break the dependency by p/invoking to the relevant OS APIs directly? I believe on Windows its a call to GetProcessTimes() and on Linux a call to getrusage(). It doesn't seem like it would take much code to do that.

Not dismissing the idea but want to point out that we have different implementations on Linux, Windows, Bsd, Mac and maybe more. @ericstj had another idea if this functionality can be exposed from Environment class (of course this will need a new API).

ericstj commented:
Pinvoking could be done via some runtime checks - you wouldn't want to introduce runtime specific builds to this component.   We try not to do that in our nuget packages.  Especially those that are used in older frameworks.

New API is blocker since that wouldn't be available on NetFx and older .NET versions.

tarekgh commented:
> New API is blocker since that wouldn't be available on NetFx and older .NET versions

I forgot about that 😢

samsp-msft commented:
Is the meter something that can be registered with DI and then fetched later, or does that create a problem with then taking a DI dependency?

tarekgh commented:
> Is the meter something that can be registered with DI and then fetched later, or does that create a problem with then taking a DI dependency?

You cannot access the DI from DiagnosticSource library level. In addition to that we need the scenario work outside containers (like console apps for instance). But you are thinking in the same direction I was thinking https://github.com/dotnet/runtime/pull/104680#issuecomment-2223691901. The Meter and instruments not having new dependencies can be created in DiagnosticSource. And in Extensions library we can start a listener which will give us the created instruments and from the instrument we can get the Meter and create more instruments that need the dependency. I know this is complicated, but it might work. The only issue I am seeing with that is we need a startup code to execute when the extensions library gets loaded.

jkotas commented:
> In theory runtime metrics should be nothing more than some calls into low-level runtime APIs (GC, JIT, assemblies, threadpool) + a tiny bit of data from the OS (some cpu and memory usage info)

Yes, I agree. The immediate problem is caused by System.Diagnostics.Process that is a heavy weight component (last time I checked, any use of it brought in several MBs of code into native aot apps). 

> New API is blocker since that wouldn't be available on NetFx and older .NET versions.

It is fine to have a solution with good performance characteristics that uses new APIs for current .NET, and a solution with worse performance characteristics that depends on pre-existing APIs for downlevel platforms and hack around the static dependency cycles using reflection.

noahfalk commented:
> It is fine to have a solution with good performance characteristics that uses new APIs for current .NET, and a solution with worse performance characteristics that depends on pre-existing APIs for downlevel platforms and hack around the static dependency cycles using reflection.

I also think it would be fine if this feature doesn't exist on downlevel platforms at all, or it doesn't show up on downlevel platforms in this release. If it weren't for the dependency on the Meter API I would have happily implemented this feature in System.Private.CoreLib where there would never be any downlevel support.

> there are more dependencies and not just Process one. Yes, process one is what is causing cycles, but I am not sure if we are ok for the other dependencies.

If we get rid of the System.Diagnostics.Process dependency that should also eliminate the System.ComponentModel.Primitives dependency. For threadpool I don't expect that is problematic as it is just a facade on S.P.C APIs. Downlevel we don't have to support the functionality. So I think Process is the only dependency that is problematic.

> Not dismissing the idea but want to point out that we have different implementations on Linux, Windows, Bsd, Mac and maybe more.

The implementation inside of System.Diagnostics.Process is more complex because it has to handle the case where the process being inspected is not the current one. When we limit to current process only the complexity is much lower. The existing cpu-usage counter calls to [RuntimeEventSourceHelper.GetCpuUsage](https://github.com/dotnet/runtime/blob/27e9b9db7a4ea10636cb28983e849c6ed1c1ea94/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs#L100) which only has two platform specific implementations:
[Unix](https://github.com/dotnet/runtime/blob/27e9b9db7a4ea10636cb28983e849c6ed1c1ea94/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSourceHelper.Unix.cs#L10) (handles Linux, BSD, Mac, and others - ultimately calls into the PAL and uses [getrusage](https://github.com/dotnet/runtime/blob/27e9b9db7a4ea10636cb28983e849c6ed1c1ea94/src/native/libs/System.Native/pal_time.c#L122))
[Windows](https://github.com/dotnet/runtime/blob/27e9b9db7a4ea10636cb28983e849c6ed1c1ea94/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSourceHelper.Windows.cs#L13) (calls to GetProcessTimes)

As best I can tell GetProcessTimes() and getrusage() covers all the cases where we'd need to make the data available, and then we've got cases like browser or WASI where the metric isn't going to be available. I'm certainly not opposed to adding a new public API, but I am guessing the quickest and easiest implementation would a conditional check at runtime to determine the platform and then p/invoke to one of those two OS APIs.

@stevejgordon - do you want to give the p/invoke approach a try for cpu.time? I believe you'd want something like:

```C#
if(OperatingSystem.IsWindows())
{
   // pinvoke to GetProcessTimes()
}
else if(!OperatingSystem.IsWasi() && !OperatingSystem.IsBrowser())
{
   // pinvoke to getrusage()
}
else
{
   // no support for cpu.time on this platform
}
```

stevejgordon commented:
@noahfalk, sure! I'll have a look at this today.

jkotas commented:
> pinvoke to getrusage

You would not want to PInvoke getrusage directly. We always want to go through our shims with stable ABIs for the reasons explained in https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/interop-guidelines.md#unix-shims .

noahfalk commented:
> You would not want to PInvoke getrusage directly

ah thanks for the heads up. I haven't tried creating a shim from scratch for a NuGet library - is that pretty straigtforward to do or you recommend we avoid it and try to create a new API in a lower level library instead?

jkotas commented:
> ah thanks for the heads up. I haven't tried creating a shim from scratch for a NuGet library

It is a pain. We only have a one library with shims (System.IO.Ports) and it is a constant source of problems. Creating a simple low-level API would be much simpler.

stevejgordon commented:
I don't know if it presents the same problems, but there is also `AppDomain.CurrentDomain.MonitoringTotalProcessorTime`, although this doesn't allow us to separate user/system time.

noahfalk commented:
> AppDomain.CurrentDomain.MonitoringTotalProcessorTime

Rats that one is close, but looking at the implementation it only reports user time and all our other cpu metrics have been user+kernel. I'd worry if we do that we'll get confusion from users that the new metric value doesn't match the older ones.

stevejgordon commented:
Would adding a new property, `AppDomain.CurrentDomain.MonitoringTotalSystemProcessorTime, ` potentially solve the issue?

noahfalk commented:
> Would proposing a new property, AppDomain.CurrentDomain.MonitoringTotalSystemProcessorTime, potentially solve the issue?

I certainly think adding a new API (such as that one, but it doesn't have to be) is a potential solution. My main concern of that approach is just whether we have time to get a new API approved within a week. @tarekgh and @ericstj probably know best there.

stephentoub commented:
> My main concern of that approach is just whether we have time to get a new API approved within a week

We do if it's simple and non-controversial.

stephentoub commented:
> AppDomain.CurrentDomain.MonitoringTotalSystemProcessorTime

AppDomain is largely deprecated, so adding new API to it isn't ideal. Is there another type on which it would make sense?

ericstj commented:
I was also thinking this could go on Environment.  It's similar to https://github.com/dotnet/runtime/issues/79146

noahfalk commented:
I'd be happy with an API like this. I'd be also fine with other names and shapes if folks don't like this one.

```C#
static class Environment
{
    // The returned ProcessCpuUsage structure should report TimeSpans that are identical to
    // calling Process.GetCurrentProcess().ProcessorUserTime and ProcessorPrivilegedTime but
    // without taking any internal dependency on the Process class.
    public static ProcessCpuUsage CpuUsage { get; }
    ...
}

public struct ProcessCpuUsage
{
    public TimeSpan UserTime { get; }
    public TimeSpan PrivilegedTime { get; }
}
```

> but then it would use Process directly in net8.0 and .NETFramework

I'd suggest we not add the new metrics to downlevel runtimes which eliminates the need for substitute dependencies.

stephentoub commented:
> The returned ProcessCpuUsage structure should report TimeSpans that are identical to
> calling Process.GetCurrentProcess().ProcessorUserTime

We should consider changing Process to use this API when it's representing the current process.

tarekgh commented:
Sounds good then. I'll create a new issue to use for the design review. 

@stephentoub @bartonjs do you have any feedback before I take this for the design review? Naming looks good to me. Alternative design would be 

```C#
static class Environment
{
    // identical tocalling Process.GetCurrentProcess().ProcessorUserTime
    public static TimeSpan ProcessorUserTime { get; }

    // identical tocalling Process.GetCurrentProcess().ProcessorPrivilegedTime
    public static TimeSpan  ProcessorPrivilegedTime { get; }
    ...
}
```

noahfalk commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/RuntimeMetrics.cs:
```diff
@@ -0,0 +1,238 @@
+
+namespace System.Diagnostics.Metrics
+{
+    internal static class RuntimeMetrics
```
I'd suggest we make this metrics only supported in 9.0+ to avoid the dependency issues targetting downlevel versions. I think a decent way to do that would be adding something like this in the PropertyGroup within the .csproj:
```
<DefineConstants Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net9.0'))">$(DefineConstants);FEATURE_RUNTIME_METRICS</DefineConstants>
```

And then making all the new code conditional on the FEATURE_RUNTIME_METRICS define. 

fyi @tarekgh in case there is a more preferred way of doing this.

tarekgh replied:
if ifdef only few lines of the code use `#if NET9_0_OR_GREATER`. If you want to ifdef the whole file, it is preferred to use something like:

```Xml
  <ItemGroup Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net9.0'))">
    <Compile Include="System\Diagnostics\Metrics\RuntimeMetrics.cs" />
  </ItemGroup>
```

noahfalk replied:
> @noahfalk are you suggesting moving all new runtime metrics to 9.0?

I'm suggesting the entire Meter be 9.0 only. Otherwise its not just the cpu.time metric that will need special-casing. A variety of the other instruments have APIs will have varying support depending on FX vs .NET 6 vs. .NET 8. I didn't want to with varying platform support on an instrument-by-instrument basis.

tarekgh replied:
This means users will continue to use OpenTelemetry instrumentation library on down-levels for the runtime metrics?

noahfalk replied:
yep! (or a few other options depending on their goals)

stevejgordon replied:
I've updated the PR to limit this to .NET 9+, as suggested. I wonder if we could support .NET 8+ instead since that's LTS? That way, consumers could grab the latest package to have almost all metrics aligned with the semantic conventions. All APIs the implementation depends on are available for .NET 8, except for the new `Environment.CpuUsage`, which could be conditionally excluded as a .NET 9+ only metric.

tarekgh commented:
I have created the issue https://github.com/dotnet/runtime/issues/104844 tracking exposing the new API.

noahfalk commented:
@stevejgordon - Hi Steve! I was chatting with @tarekgh and since time is drawing short to get these PRs merged here is what we believed would work best. Does this work for you?
1. The new CPU usage API was approved earlier today (https://github.com/dotnet/runtime/issues/104844#issuecomment-2231556034)
2. Tarek will prepare a PR that implements the new API tomorrow, and if possible merge it tomorrow.
3. In parallel tomorrow you could prepare the runtime metrics PR with the cpu.time metric commented out. I'm only aware of [one outstanding request] (https://github.com/dotnet/runtime/pull/104680/files#r1676553244) on your current PR.
4. If Tarek gets his PR merged tomorrow you can update yours to consume it and hopefully merge on Thursday. If Tarek doesn't get his PR merged tomorrow we can merge your runtime metrics without the cpu.time instrument included and then Tarek can include cpu.time in his own PR that merges later this week.

Originally I believed 7/18 was checkin deadline but apparently I was off by a day, 7/19 is our deadline.

stevejgordon committed:
PR feedback and naming updates

- Limits the new meter to .NET 9+ per the review feedback.
- Comment out dotnet.process.cpu.time until blocking PR for a new API is merged.
- Renames several counters to align with latest spec.

stevejgordon commented:
> @stevejgordon - Hi Steve! I was chatting with @tarekgh and since time is drawing short to get these PRs merged here is what we believed would work best. Does this work for you?
> 
> 1. The new CPU usage API was approved earlier today ([Introducing Environment.ProcessorUserTime & ProcessorPrivilegedTime  #104844 (comment)](https://github.com/dotnet/runtime/issues/104844#issuecomment-2231556034))
> 2. Tarek will prepare a PR that implements the new API tomorrow, and if possible merge it tomorrow.
> 3. In parallel tomorrow you could prepare the runtime metrics PR with the cpu.time metric commented out. I'm only aware of [one outstanding request](https://github.com/dotnet/runtime/pull/104680/files#r1676553244) on your current PR.
> 4. If Tarek gets his PR merged tomorrow you can update yours to consume it and hopefully merge on Thursday. If Tarek doesn't get his PR merged tomorrow we can merge your runtime metrics without the cpu.time instrument included and then Tarek can include cpu.time in his own PR that merges later this week.
> 
> Originally I believed 7/18 was checkin deadline but apparently I was off by a day, 7/19 is our deadline.

@noahfalk, sounds good. I've updated this as suggested, including the commented-out `dotnet.process.cpu.time` metric.

noahfalk commented:
It looks like some of the new tests in this PR are also failing in CI: https://dev.azure.com/dnceng-public/public/_build/results?buildId=744347&view=ms.vss-test-web.build-test-results-tab&runId=18807774&resultId=166339&paneView=debug

stevejgordon commented:
> It looks like some of the new tests in this PR are also failing in CI: https://dev.azure.com/dnceng-public/public/_build/results?buildId=744347&view=ms.vss-test-web.build-test-results-tab&runId=18807774&resultId=166339&paneView=debug

@noahfalk - Ah, okay. I see one obvious issue with the `ExceptionsCount` test in `RuntimeMetricsTests`, which I'll fix. For those failing in that particular output, I wonder if the GC APIs return different results on ARM. For example, the test for `dotnet.gc.pause.time` validates the value is greater than zero, but this fails on this platform. I quickly looked at the GC code and didn't spot an obvious reason why this might come back < 0. I'll dig a bit deeper. I've yet to look at the others, but I will focus on those later this morning.

I was sure I'd run `MetricsTests` also, but apparently not. I'll look at those ones, as they seem to be picking up the new instruments, so they may need some tweaking to limit what they are asserting against.

stevejgordon committed:
Fix exceptions count test

noahfalk commented:
> I quickly looked at the GC code and didn't spot an obvious reason why this might come back < 0. I'll dig a bit deeper.

If it helps, feel free to add logging to a test case and then update the PR causing it to run. It might be that there is some pre-existing bug with that API on ARM that has gone unnoticed.

stevejgordon committed:
Add some test logging

stevejgordon committed:
Fix MetricsTests

Ensure that each test only observes the expected Meter

stevejgordon committed:
More test logging and extended wait for metrics

stevejgordon committed:
More logging and a test fix

stevejgordon committed:
Remove System.Diagnostics.Process

stevejgordon commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj:
```diff
@@ -126,15 +126,21 @@ System.Diagnostics.DiagnosticSource</PackageDescription>
   <ItemGroup Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)'">
     <Reference Include="System.Collections" />
     <Reference Include="System.Collections.Concurrent" />
+    <Reference Include="System.Diagnostics.Process" />
```
Yep!

stevejgordon committed:
Update HeapTags test to assert correctly if no GC has run

stevejgordon committed:
Only include System.Threading.ThreadPool reference for net9.0

stevejgordon commented:
@noahfalk, I have tweaked the tests to accommodate some of the platforms not running a forced GC. I also tweaked the existing tests that were failing to ensure they only observe their expected `Meter`s. I'm hoping this current run should go through on all the tests, but I will keep an eye on it.

tarekgh committed:
Merge branch 'main' into runtime-metrics

tarekgh commented:
@stevejgordon I have changed the PR to be not draft. I think we can merge once we get the green CI and no more comments need to be addressed. I can enable the cpu time metrics with my changes later as I need more time to finish exposing the new Environment API.

tarekgh commented:
@stevejgordon there are some tests failing on WASM 

- https://[helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-104680-merge-0b04ac1b17ef47dfa6/WasmTestOnChrome-ST-System.Diagnostics.DiagnosticSource.Tests/1/console.7b3e2b0f.log?helixlogtype=result](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-104680-merge-0b04ac1b17ef47dfa6/WasmTestOnChrome-ST-System.Diagnostics.DiagnosticSource.Tests/1/console.7b3e2b0f.log?helixlogtype=result)
- https://[helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-104680-merge-0b04ac1b17ef47dfa6/WasmTestOnChrome-ST-System.Diagnostics.DiagnosticSource.Tests/1/console.7b3e2b0f.log?helixlogtype=result](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-104680-merge-0b04ac1b17ef47dfa6/WasmTestOnChrome-ST-System.Diagnostics.DiagnosticSource.Tests/1/console.7b3e2b0f.log?helixlogtype=result)

Maybe we need to exclude running the metric tests on WASM. something like `[ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotBrowser))]`. We already do that in other metric tests.

stevejgordon commented:
@tarekgh, sure. Will try that first thing tomorrow. I was expecting something like that might be needed somewhere.

noahfalk commented:
Just a heads up, I lost internet connectivity at my house. I'm doing what I can from my cell phone but it's a subpar UI experience to say the least :)

stevejgordon committed:
Skip tests on browser

stevejgordon commented:
@noahfalk / @tarekgh It looks like the tests are happy now! The only failures seem to be known issues and unrelated.

tarekgh commented:
Thanks @stevejgordon for your contribution.

stevejgordon commented:
Happy to help! Thanks for the support as we got this over the line!

noahfalk commented:
Thanks Steve!
